proposals/AmbiguousType-pragma.rst

```diff
@@ -0,0 +1,242 @@
+
+Proposed Change Specification
+-----------------------------
+
+1. There is to be a pragma ``{-# AMBIGUOUS #-}``, to appear immediately after the ``::`` (so before the type), in principle wherever ``::`` can appear. That is a function or method definition's signature; term type annotations beginning ``::``; pattern signatures (and pattern synonyms). The "in principle" would also include ``::`` introducing kind signatures (type of types); and possibly places ready for not-yet-developed usages of type ``@`` applications. Examples::
+
+        f :: {-# AMBIGUOUS #-} C a => Int
+
+        class Sized a  where
+          sizeOf :: {-# AMBIGUOUS #-} Integer
+      
+        apply (x :: {-# AMBIGUOUS #-} forall a. (Read a, Show a) => String -> String) = x @Int "01"
+    
+        norm :: {-# AMBIGUOUS #-} forall a. (Read a, Show a) => String -> String
+        norm = show @a . read                                     -- needs type-lambda #155
+    
+        data T :: {-# AMBIGUOUS #-} F a -> Type                   -- ambiguous kind signature
+        
+   Also ``{-# AMBIGUOUS #-}`` can appear after ``::`` in a ``data`` or ``newtype`` declaration -- either for fields or GADT-style constructors; or in H2010-style immediately ~after~ before the constructor (that is, after ``=`` or ``|``, because no ``::``)::
+   
+        data D a where                                            -- GADT style
+          MkD :: {-# AMBIGUOUS #-} (forall b. C a b => a)
+
+        data D2 a = {-# AMBIGUOUS #-} MkD2  (forall b. C a b => a) (Show a => a) 
+                                               -- H2010 style; note no ::; note the second arg is not ambiguous
+                                               
+        newtype N a = {-# AMBIGUOUS #-} MkN { foo :: {-# AMBIGUOUS #-} (forall b. C a b => a) } 
+                                               -- AMBIGUOUS not ambiguous, but one is redundant
+
+   
+   See Appendix giving a diff to the BNF for the language (wrt the Report, and affected extensions).
+
+2. Signatures marked ``AMBIGUOUS`` are to be validated as if ``-XAllowAmbiguousTypes`` is set, for that signature only. (If that is already set module-wide, the pragma has the effect of suppressing the ``-Wambiguous-type`` warning, see 5.)
+
+3. This does not change the validation for ambiguous types/type variables at usage sites.
+
+4. The error reporting from the ambiguity check that currently suggests ``To defer the ambiguity check to use sites, enable AllowAmbiguousTypes`` must make clear this is likely to entail using ``TypeApplications`` at usage sites, and that there are several possible approaches to avoid ambiguous type variables.
+
+   Precise wording to be arrived at in discussion of this PR. (Prefer not mentioning ``AllowAmbiguousTypes`` at all.) Starting bikeshed::
```

Comment:
This sentence should be removed before merge.

My attempt:

> If you intend to resolve the ambiguity at each use site (typically via TypeApplications), prefix this signature with the AMBIGUOUS pragma.

Reply:
Done. Thank you.